ImageFeed/Auth/OAuth2Service.swift

```diff
@@ -5,48 +5,86 @@ class OAuth2Service {
     private enum NetworkError: Error {
         case codeError
     }
+    private let urlSession = URLSession.shared
+    private var task: URLSessionTask?
+    private var lastCode: String?
+    
+    private func makeRequest(code: String) -> URLRequest {
+        var urlComponents = URLComponents(string: Constants.oauthString)!
+        urlComponents.queryItems = [
+            URLQueryItem(name: "client_id", value: Constants.accessKey),
+            URLQueryItem(name: "client_secret", value: Constants.secretKey),
+            URLQueryItem(name: "redirect_uri", value: Constants.redirectUri),
+            URLQueryItem(name: "code", value: code),
+            URLQueryItem(name: "grant_type", value: "authorization_code")
+        ]
+        
+        guard let url = urlComponents.url else {
+            fatalError("makeRequest Error")
+        }
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        return request
+    }
+    
     func fetchAuthToken(
         code: String,
-        handler: @escaping (Result<String, Error>) -> Void) {
-            var urlComponents = URLComponents(string: Constants.oauthString)!
-            urlComponents.queryItems = [
-                URLQueryItem(name: "client_id", value: Constants.accessKey),
-                URLQueryItem(name: "client_secret", value: Constants.secretKey),
-                URLQueryItem(name: "redirect_uri", value: Constants.redirectUri),
-                URLQueryItem(name: "code", value: code),
-                URLQueryItem(name: "grant_type", value: "authorization_code")
-            ]
-            let url = urlComponents.url!
+        handler: @escaping (Result<OAuthTokenResponseBody, Error>) -> Void) {
             
-            var request = URLRequest(url: url)
-            request.httpMethod = "POST"
+            assert(Thread.isMainThread)
+            if lastCode == code { return }
```

Comment:
такой код проверяет на валидность состояние запроса, поэтому конструкция guard больше подходит :) но на вкус и цвет, как говорится
```suggestion
            guard lastCode != code else { return }
```